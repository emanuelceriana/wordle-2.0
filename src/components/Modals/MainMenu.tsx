import { memo, useContext } from "react";
import { AppContext, IAppContext } from "../../context/AppContext";
import { RangeSlider } from "../Selectors/RangeSlider";
import { WordLengthSelector } from "../Selectors/WordLengthSelector";
import { defaultTriesCount, defaultWordLength } from "../Selectors/constants";
import styles from "./MainMenu.module.scss";
import { Modal } from "./Modal";
import { mainMenuText } from "./constants";

export const MainMenu = memo(() => {
  const { isMainMenuOpen, setWordLength, setTriesCount, startGame } =
    useContext<IAppContext>(AppContext);

  return (
    <Modal isOpen={isMainMenuOpen} showCloseButton={false}>
      <div className={styles.mainMenu}>
        <h3>Welcome to Wordle 2.0</h3>
        <p>{mainMenuText}</p>

        <WordLengthSelector
          defaultValue={defaultWordLength}
          onClickRequest={setWordLength}
        />
        <RangeSlider
          defaultValue={defaultTriesCount}
          onChangeRequest={setTriesCount}
        />
        <div>
          <button onClick={startGame}>Start Game</button>
        </div>
      </div>
    </Modal>
  );
});
