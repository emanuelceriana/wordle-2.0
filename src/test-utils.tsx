import React, { ReactNode } from "react";
import { render } from "@testing-library/react";
import { AppContext } from "./context/AppContext";

const MockProvider: React.FC<{
  children: ReactNode | ReactNode[];
  value: object;
}> = ({ children, value }) => (
  <AppContext.Provider value={value}>{children}</AppContext.Provider>
);

export function renderWithContext(element: React.ReactElement, value: object) {
  return render(<MockProvider value={value}>{element}</MockProvider>);
}

export function withContext(element: React.ReactElement, value: object) {
  return <MockProvider value={value}>{element}</MockProvider>;
}
