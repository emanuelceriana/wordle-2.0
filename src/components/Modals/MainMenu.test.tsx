import { act, fireEvent } from "@testing-library/react";
import { MainMenu } from "./MainMenu";
import { renderWithContext } from "../../test-utils";

interface MockValueContext {
  isMainMenuOpen: boolean;
  setWordLength: () => void;
  setTriesCount: () => void;
  startGame: () => void;
}

describe("<MainMenu />", () => {
  const startGameMock = jest.fn();

  let mockValueContext: MockValueContext = {
    isMainMenuOpen: true,
    setWordLength: () => {},
    setTriesCount: () => {},
    startGame: startGameMock,
  };

  it("should render without errors", () => {
    const { asFragment } = renderWithContext(<MainMenu />, mockValueContext);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should be closed if mainMenu is false", () => {
    mockValueContext.isMainMenuOpen = false;

    const { queryByTestId } = renderWithContext(<MainMenu />, mockValueContext);

    const mainMenu = queryByTestId("mainMenu");

    expect(mainMenu).toBe(null);
  });
  it("should be open if helMenuOpen is true", () => {
    mockValueContext.isMainMenuOpen = true;

    const { getByTestId } = renderWithContext(<MainMenu />, mockValueContext);

    const mainMenu = getByTestId("mainMenu");

    expect(mainMenu).toBeDefined();
  });
  it("should call startGame when Start Game button is clicked", () => {
    const { getByText } = renderWithContext(<MainMenu />, mockValueContext);

    const startGameButton = getByText("Start Game");

    act(() => {
      fireEvent.click(startGameButton);
    });

    expect(mockValueContext.startGame).toHaveBeenCalledTimes(1);
  });
});
