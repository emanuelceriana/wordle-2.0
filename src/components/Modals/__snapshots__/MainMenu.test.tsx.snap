// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainMenu /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="modalBg"
  />
  <div
    class="modalBg__modal"
  >
    <div
      class="modalBg__modal-title"
    />
    <div
      class="modalBg__modal-body"
    >
      <div
        class="mainMenu"
        data-testid="mainMenu"
      >
        <h3>
          Welcome to Wordle 2.0
        </h3>
        <p>
          The well known game but with an awesome feature
You can select the length of the word to guess and the amount of tries!
        </p>
        <div>
          <div
            class="wordLengthSelector"
          >
            <span>
              Word Length:
            </span>
            <div
              class="wordLengthSelector__buttons"
            >
              <button
                class="wordLengthSelector__button"
              >
                4
              </button>
              <button
                class="wordLengthSelector__button wordLengthSelector__selected"
              >
                5
              </button>
              <button
                class="wordLengthSelector__button"
              >
                6
              </button>
              <button
                class="wordLengthSelector__button"
              >
                7
              </button>
              <button
                class="wordLengthSelector__button"
              >
                8
              </button>
              <button
                class="wordLengthSelector__button"
              >
                9
              </button>
            </div>
          </div>
        </div>
        <div
          class="rangeSlider"
        >
          <span
            class="rangeSlider__label"
          >
            Try Count:
          </span>
          <div
            class="rangeSlider__input"
          >
            <span>
              1
            </span>
            <input
              max="50"
              min="1"
              step="1"
              type="range"
              value="6"
            />
            <span>
              50
            </span>
            <div
              class="rangeSlider__counter"
            >
              6
            </div>
          </div>
        </div>
        <div>
          <button>
            Start Game
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
